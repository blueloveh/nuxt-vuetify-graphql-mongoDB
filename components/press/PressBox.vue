<script>
import sample from '@/components/plugins/Sample'

export default {
  functional: true,
  name: 'PressBox',
  props: {
    type: {
      type: String,
      require: true,
      default: ''
    },
    label: {
      type: String,
      require: true,
      default: ''
    },
    nodes: {
      type: Array,
      require: true,
      default: null,
      useDefaultForNull: true
    }
  },
  render(h, { props, listeners, slots, data }) {
    function createButton(item) {
      return h(
        'v-toolbar',
        {
          attrs: {
            dense: false,
            floating: false
          }
        },
        [
          h('v-toolbar-title', [`${item.label}`]),
          h('v-spacer'),
          h('v-btn', { attrs: { icon: true } }, ['dk'])
        ]
      )
    }

    function createContainer(item) {
      return h(
        'v-container',
        {
          class: {
            grey: true
          },
          attrs: {
            fluid: true
          }
        },
        [createButton(item), ...createNodes(item.nodes, 'layout')]
      )
    }

    function createLayout(item) {
      return h(
        'v-layout',
        {
          class: {
            red: true
          },
          attrs: {
            row: true,
            wrap: true,
            xs12: true,
            'align-center': true,
            'justify-center': true
          }
        },
        item.nodes
          ? [createButton(item), ...createNodes(item.nodes, 'flex')]
          : [createButton(item), `${item.type} - ${item.label}`]
      )
    }

    function createFlex(item) {
      return h(
        'v-flex',
        {
          class: {
            yellow: true
          },
          attrs: {
            xs12: true,
            sm6: true,
            md3: true,
            'align-center': true,
            'justify-center': true
          }
        },
        item.comp
          ? [createButton(item), h(sample)]
          : [createButton(item), `${item.type} - ${item.label}`]
      )
    }

    function createCard(item) {
      return h(
        'v-card',
        {
          class: {
            blue: true
          },
          attrs: {
            xs12: true,
            'align-center': true,
            'justify-center': true
          }
        },
        [createButton(item), `${item.type} - ${item.label}`]
      )
    }

    function chooseNode(nodeName, item) {
      let node = null
      switch (nodeName) {
        case 'layout':
          node = createLayout(item)
          break
        case 'flex':
          node = createFlex(item)
          break
        case 'card':
          node = createCard(item)
          break
        default:
          node = createContainer(item)
      }
      return node
    }

    function createNodes(nodes, nodeName) {
      return nodes.map(function(item) {
        return item.nodes ? createScopedSlot(item) : chooseNode(item.type, item)
      })
    }

    function createScopedSlot(item) {
      return data.scopedSlots.default({
        nodes: item.nodes,
        label: item.label,
        type: item.type
      })
    }

    let grid = null
    if (props.type === 'container') {
      console.log('container!!')
      grid = createContainer(props)
    } else if (props.type === 'layout') {
      console.log('layout!!')
      grid = createLayout(props)
    } else if (props.type === 'flex') {
      console.log('flex!!')
      grid = createFlex(props)
    } else {
      console.log('card!!')
      grid = createCard(props)
    }
    return grid
  }
}
</script>

<style scoped>
.layout {
  min-height: 100px;
}
.flex {
  min-height: 80px;
}
.v-card {
  min-height: 60px;
}
.fab-container {
  position: fixed;
  bottom: 0;
  right: 0;
}
</style>
