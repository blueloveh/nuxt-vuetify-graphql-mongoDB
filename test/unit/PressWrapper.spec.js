import { mount } from '@vue/test-utils'
import PressBox from '@/components/press/PressBox.vue'
import Vuetify from 'vuetify'
import Vue from 'vue'

Vue.use(Vuetify)

describe('PressBox', () => {
  test('makes two layouts', () => {
    const wrapper = mount(PressBox, {
      context: {
        props: {
          label: 'page',
          type: 'container',
          nodes: [
            {
              label: 'layout1',
              type: 'layout',
              nodes: [
                {
                  label: 'flex1.1',
                  type: 'flex'
                },
                {
                  label: 'flex1.2',
                  type: 'flex',
                  comp: 'helloWorld'
                }
              ]
            },
            {
              label: 'layout2',
              type: 'layout'
            }
          ]
        }
      },
      scopedSlots: {
        default: someProp => {
          const scoped = mount(PressBox, {
            context: { props: { ...someProp } }
          })
          return scoped.vnode
        }
      }
    })

    expect(wrapper.element).toMatchSnapshot()

    const layoutArray = wrapper.findAll('.flex')
    expect(layoutArray.length).toBe(1)
  })
})
