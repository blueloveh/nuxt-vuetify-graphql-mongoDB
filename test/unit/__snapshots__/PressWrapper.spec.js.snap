// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PressBox makes two layouts 1`] = `
<div
  class="container fluid grey"
>
  <nav
    class="v-toolbar theme--light"
    style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);"
  >
    <div
      class="v-toolbar__content"
      style="height: 64px;"
    >
      <div
        class="v-toolbar__title"
      >
        page
      </div>
      <div
        class="spacer"
      />
      <button
        class="v-btn v-btn--icon theme--light"
        type="button"
      >
        <div
          class="v-btn__content"
        >
          dk
        </div>
      </button>
    </div>
  </nav>
  <div
    class="layout row wrap xs12 align-center justify-center red"
  >
    <nav
      class="v-toolbar theme--light"
      style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);"
    >
      <div
        class="v-toolbar__content"
        style="height: 64px;"
      >
        <div
          class="v-toolbar__title"
        >
          layout1
        </div>
        <div
          class="spacer"
        />
        <button
          class="v-btn v-btn--icon theme--light"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            dk
          </div>
        </button>
      </div>
    </nav>
    <div
      class="flex xs12 sm6 md3 align-center justify-center yellow"
    >
      <nav
        class="v-toolbar theme--light"
        style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);"
      >
        <div
          class="v-toolbar__content"
          style="height: 64px;"
        >
          <div
            class="v-toolbar__title"
          >
            flex1.1
          </div>
          <div
            class="spacer"
          />
          <button
            class="v-btn v-btn--icon theme--light"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              dk
            </div>
          </button>
        </div>
      </nav>
      flex - flex1.1
    </div>
    <div
      class="flex xs12 sm6 md3 align-center justify-center yellow"
    >
      <nav
        class="v-toolbar theme--light"
        style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);"
      >
        <div
          class="v-toolbar__content"
          style="height: 64px;"
        >
          <div
            class="v-toolbar__title"
          >
            flex1.2
          </div>
          <div
            class="spacer"
          />
          <button
            class="v-btn v-btn--icon theme--light"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              dk
            </div>
          </button>
        </div>
      </nav>
      <div>
        <h1>
          Hello World!
        </h1>
         
        <ul>
          <li>
            1
          </li>
          <li>
            2
          </li>
          <li>
            3
          </li>
          <li>
            4
          </li>
          <li>
            5
          </li>
          <li>
            6
          </li>
          <li>
            7
          </li>
          <li>
            8
          </li>
          <li>
            9
          </li>
          <li>
            10
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="layout row wrap xs12 align-center justify-center red"
  >
    <nav
      class="v-toolbar theme--light"
      style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);"
    >
      <div
        class="v-toolbar__content"
        style="height: 64px;"
      >
        <div
          class="v-toolbar__title"
        >
          layout2
        </div>
        <div
          class="spacer"
        />
        <button
          class="v-btn v-btn--icon theme--light"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            dk
          </div>
        </button>
      </div>
    </nav>
    layout - layout2
  </div>
</div>
`;
